### PR TITLE
Use psrpc

### PR DESCRIPTION
**Pull request by frostbyte73**

### CODE REVIEW

File: cmd/server/main.go [hunk 1]
```diff
@@ -83,13 +90,28 @@ func runService(c *cli.Context) error {
...
-	rc, err := redis.GetRedisClient(conf.Redis)
+	rc, err := lkredis.GetRedisClient(conf.Redis)
 	if err != nil {
 		return err
 	}
 
-	rpcServer := egress.NewRedisRPCServer(rc)
-	svc := service.NewService(conf, rpcServer)
+	ctx := context.Background()
+	previous, err := rc.Get(ctx, ev.EgressVersionKey).Result()
```
Comment (davidzhao) [inline cmd/server/main.go:99]: what happens when multiple egress instances are running? like as part of a rolling deploy?

File: pkg/service/service.go [hunk 2]
```diff
@@ -103,104 +113,71 @@ func (s *Service) Run() error {
...
+func (s *Service) StartEgressAffinity(req *livekit.StartEgressRequest) float32 {
+	if !s.manager.canAccept(req) || !s.monitor.CanAcceptRequest(req) {
+		return 0
+	}
 
-		s.sendResponse(ctx, req, p.Info, err)
-		if err != nil {
-			span.RecordError(err)
-		}
+	if s.manager.isIdle() {
```
Comment (davidzhao) [inline pkg/service/service.go:141]: what do these constants mean? comments here would be helpful.

File: pkg/service/service.go [hunk 3]
```diff
@@ -14,34 +14,44 @@ import (
...
+		monitor:     monitor,
+		manager:     NewProcessManager(conf, monitor),
+		shutdown:    make(chan struct{}),
 	}
+
+	psrpcServer, err := rpc.NewEgressInternalServer(conf.NodeID, s, bus)
+	if err != nil {
+		return nil, err
+	}
+	s.psrpcServer = psrpcServer
```
Comment (davidzhao) [inline pkg/service/service.go:53]: how is this used? I don't see references to psrpcServer anywhere.. does it just get started automatically?

File: test/integration.go [hunk 4]
```diff
@@ -377,7 +404,13 @@ func runSegmentsTest(t *testing.T, conf *TestConfig, req *livekit.StartEgressReq
 
 func startEgress(t *testing.T, conf *TestConfig, req *livekit.StartEgressRequest) string {
 	// send start request
-	info, err := conf.rpcClient.SendRequest(context.Background(), req)
+	var info *livekit.EgressInfo
+	var err error
+	if conf.PSRPC {
+		info, err = conf.psrpcClient.StartEgress(context.Background(), req)
```
Comment (davidzhao) [inline test/integration.go:410]: nice perk of having matching method names.

File: pkg/service/service.go (hunk 3 above)
Comment (frostbyte73) [reply pkg/service/service.go:53]: added a shutdown method to drain and close

Comment (davidzhao) [reply pkg/service/service.go:53]: so it does start automatically when the server is created?

Comment (frostbyte73) [reply pkg/service/service.go:53]: yeah

**Review by davidzhao — APPROVED** [review]

File: cmd/server/main.go (hunk 1 above)
Comment (frostbyte73) [reply cmd/server/main.go:99]: I think this should be ok with redis? I could add locking, but for that we just use a SetNX and check which one got the write

**Pull request merged**